app/Main.hs

```diff
@@ -32,57 +42,114 @@ main :: IO ()
 main = do
     hSetBuffering stdin  LineBuffering
     hSetBuffering stdout LineBuffering
-    environment <- Env <$> T.getContents
-                       <*> defaultConfig
+    environment <- Env <$> defaultConfig
     inputFromTerminal <- hIsTerminalDevice stdin
     outputToTerminal  <- hIsTerminalDevice stdout
     args <- getArgs
-    runVgrepT environment $ case (inputFromTerminal, outputToTerminal) of
-        (True,  False)  -> recursiveGrep >>= liftIO . T.putStrLn
-        (False, False)  -> grepStdin     >>= liftIO . T.putStrLn
+    case (inputFromTerminal, outputToTerminal) of
+        (True,  False)  -> runHeadless (const recursiveGrep)
+        (False, False)  -> runHeadless grep
         (False, True)
-            | null args -> view input      >>= runApp_ . app
-            | otherwise -> grepStdinForApp >>= runApp_ . app
-        (True,  True)   -> recursiveGrep   >>= runApp_ . app
+            | null args -> runGui environment id
+            | otherwise -> runGui environment grepForApp
+        (True,  True)   -> runGui environment (const recursiveGrep)
+  where
+    stdinText  = P.stdinLn  >-> P.map T.pack
+    stdoutText = P.stdoutLn <-< P.map T.unpack
+    runHeadless grepCommand = runEffect (grepCommand stdinText >-> stdoutText)
+    runGui environment  grepCommand = withSpawn unbounded $
+      \(evSink, evSource) -> do
+        let stdinText' = stdinText >-> P.tee (P.map ReceiveInputEvent >-> toOutput evSink)
+        grepThread <- async . runEffect $
+            grepCommand stdinText' >-> P.map ReceiveResultEvent
+                                   >-> toOutput evSink
+        runApp_ app environment (fromInput evSource)
+        cancel grepThread
 
 
 type MainWidget = HSplitWidget ResultsWidget PagerWidget
 
-app :: Text -> App MainWidget
-app grepOutput = App
-    { _initialize  = initSplitView
-    , _handleEvent = eventHandler
-    , _render      = fmap picForImage . drawWidget }
+data AppState = AppState { _appWidget  :: MainWidget
+                         , _inputLines :: Seq Text }
+
+appWidget :: Lens' AppState MainWidget
+appWidget = lens _appWidget (\s w -> s { _appWidget = w })
+
+inputLines :: Lens' AppState (Seq Text)
+inputLines = lens _inputLines (\s l -> s { _inputLines = l })
+
+data Event = VtyEvent Vty.Event
+           | ReceiveInputEvent  Text
+           | ReceiveResultEvent Text
+
+vtyEvent :: Prism' Event Vty.Event
```

Comment:
Good question. Wanted to see the mechanics behind `makePrisms` and `makeLenses`, and it didn't hurt enough to change it afterwards.